### PR TITLE
Encoding for the ncdata "shortcut" datapaths

### LINKED ISSUE #255: Incompatible with iris 3.16
It looks like `ncdata` v0.3.2 is not compatible with the latest iris release, v3.16.0 (tested with xarray 2026.7.0).

```
.pixi/envs/default/lib/python3.14/site-packages/ncdata/iris_xarray.py:54: in cubes_from_xarray
    cubes = to_iris(ncdata, **iris_load_kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.pixi/envs/default/lib/python3.14/site-packages/ncdata/iris.py:53: in to_iris
    cubes = CubeList(iris.load(dslikes, **iris_load_kwargs))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.pixi/envs/default/lib/python3.14/site-packages/iris/loading.py:235: in load
    cubes = _load_collection(uris, constraints, callback).combined().cubes()
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.pixi/envs/default/lib/python3.14/site-packages/iris/loading.py:203: in _load_collection
    result = _CubeFilterCollection.from_cubes(cubes, constraints)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.pixi/envs/default/lib/python3.14/site-packages/iris/loading.py:159: in from_cubes
    for c in cubes:
             ^^^^^
.pixi/envs/default/lib/python3.14/site-packages/iris/loading.py:108: in _generate_cubes
    for cube in iris.io.load_data_objects(data_objects, callback):
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.pixi/envs/default/lib/python3.14/site-packages/iris/io/__init__.py:306: in load_data_objects
    yield from load_http(urls, callback)
.pixi/envs/default/lib/python3.14/site-packages/iris/io/__init__.py:286: in load_http
    for cube in handling_format_spec.handler(fnames, callback):
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.pixi/envs/default/lib/python3.14/site-packages/iris/fileformats/netcdf/loader.py:712: in load_cubes
    with CFReader(file_source) as cf:
         ^^^^^^^^^^^^^^^^^^^^^
.pixi/envs/default/lib/python3.14/site-packages/iris/fileformats/cf.py:1356: in __init__
    ds.set_auto_chartostring(False)
    ^^^^^^^^^^^^^^^^^^^^^^^^
.pixi/envs/default/lib/python3.14/site-packages/ncdata/dataset_like.py:74: in __getattr__
    return self.getncattr(attr)
           ^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ncdata.dataset_like.Nc4DatasetLike object at 0x7c1cf981dd30>, attr = 'set_auto_chartostring'

    def getncattr(self, attr: str):
        attrs = self._ncdata.avals
        if attr in attrs:
            result = attrs[attr]
        else:
            # Don't allow it to issue a KeyError, as this upsets 'getattr' usage.
            # Raise an AttributeError instead.
>           raise AttributeError(attr)
E           AttributeError: set_auto_chartostring

.pixi/envs/default/lib/python3.14/site-packages/ncdata/dataset_like.py:59: AttributeError
```

I noticed you have some work in progress to fix this in #253 and #254

### LINKED ISSUE #7264: Scalar netcdf character variable, not loading with Iris 3.16
## How To Reproduce
Steps to reproduce the behaviour:
 run in Iris 3.16 ..

```
import netCDF4 as nc
ds = nc.Dataset("tmp.nc", "w")
v = ds.createVariable("vx", 'S1')
v[:] = b'x'
ds.close()

cubes = iris.load("tmp.nc")
print(cubes)
c = cubes[0]
print(c)
print(repr(c.data))
```

## Expected behaviour
output
```
0: vx / (unknown)                      (scalar cube)
vx / (unknown)                      (scalar cube)
array('', dtype='<U1')
```
## Actual 
```
Traceback (most recent call last):
  File "/home/users/patrick.peglar/worklog/period_20260824_zarr/support/dimitar_vlaykov_iris3.16_load_problem/example.py", line 17, in <module>
    cubes = iris.load("tmp.nc")
            ^^^^^^^^^^^^^^^^^^^
  File "/home/users/patrick.peglar/git/iris/lib/iris/loading.py", line 235, in load
    cubes = _load_collection(uris, constraints, callback).combined().cubes()
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/patrick.peglar/git/iris/lib/iris/loading.py", line 203, in _load_collection
    result = _CubeFilterCollection.from_cubes(cubes, constraints)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/patrick.peglar/git/iris/lib/iris/loading.py", line 159, in from_cubes
    for c in cubes:
             ^^^^^
  File "/home/users/patrick.peglar/git/iris/lib/iris/loading.py", line 100, in _generate_cubes
    for cube in iris.io.load_files(part_names, callback, constraints):
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/patrick.peglar/git/iris/lib/iris/io/__init__.py", line 249, in load_files
    for cube in handling_format_spec.handler(fnames, callback, constraints):
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/patrick.peglar/git/iris/lib/iris/fileformats/netcdf/loader.py", line 774, in load_cubes
    cube = _load_cube(engine, cf, cf_var, cf.filename)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/patrick.peglar/git/iris/lib/iris/fileformats/netcdf/loader.py", line 409, in _load_cube
    return _load_cube_inner(engine, cf, cf_var, filename)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/patrick.peglar/git/iris/lib/iris/fileformats/netcdf/loader.py", line 424, in _load_cube_inner
    data = _get_cf_var_data(cf_var)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/patrick.peglar/git/iris/lib/iris/fileformats/netcdf/loader.py", line 283, in _get_cf_var_data
    result = cf_var[:]
             ~~~~~~^^^
  File "/home/users/patrick.peglar/git/iris/lib/iris/fileformats/cf.py", line 209, in __getitem__
    return self.cf_data.__getitem__(key)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/patrick.peglar/git/iris/lib/iris/fileformats/netcdf/_bytecoding_datasets.py", line 383, in __getitem__
    data = encoding_spec.decode_bytes_to_stringarray(data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/patrick.peglar/git/iris/lib/iris/fileformats/netcdf/_bytecoding_datasets.py", line 196, in decode_bytes_to_stringarray
    data = decode_bytesarray_to_stringarray(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/users/patrick.peglar/git/iris/lib/iris/fileformats/netcdf/_bytecoding_datasets.py", line 80, in decode_bytesarray_to_stringarray
    bytes = b"".join([b or b"\0" for b in element_bytes])
                                          ^^^^^^^^^^^^^
TypeError: 'numpy.float64' object is not iterable
```

## Context
From some investigation, this is a specific error with scalar char variables.
A numeric variable does not have this problem, 
neither does a character variable with  dimensions / a shape

### LINKED ISSUE #7283: ncdata broken by Iris 3.16
See : https://github.com/SciTools/ncdata/issues/255

Most obviously, ncdata must add a 'set_auto_stringtochar'.  To be fixed in ncdata (cf https://github.com/SciTools/ncdata/pull/253).
But more subtly, the 'shortcut' methods are missing the new encode/decode support.
#7281 fixes that

### PR DESCRIPTION
**Pull request by pp-mo**
Closes #7264

but mostly, fixes https://github.com/SciTools/ncdata/issues/255
( although..
 [this change](https://github.com/SciTools/ncdata/pull/253/commits/72e285e38757816935e0251f2691291b2e007686) or equivalent 
will also be essential for successful interoperation with Iris 3.16
)

This solves the problems introduced in v3.16.0 
..because.. the "bypass" mechanism for loading+saving to/from ncdata "like a file" is unfortunately **_also_** bypassing the encode/decode operations.

### Implementation notes : 
  * it is not practical to wrap an ncdata NcVariableLike in an bytecoding_datasets.EncodedVariable, since this can cause hanging when attempting to access data : because, the EncodedVariable acquires the file-system lock before calling ncdata, but ncdata then also wants to acquire the lock to read data.
  * ... i.e. the encode/decode is intrinsically tied to the "threadsafe-datasets" operation, given how it is implemented.
  * .. so instead, I mirrored the encode/decode operation provided by EncodedDataset, within the "shortcut" operations which are triggered when a variable has the "._data_array" property.
  * **however** it also means that encode/decode need to work on lazy data.  
     So I've fixed that

## TODO:
  * [x] more specific tests for new code (and possibly, VariableEncoder generally ??)
  * [x] whatsnew

### CODE REVIEW

File: lib/iris/tests/unit/fileformats/netcdf/loader/test__get_cf_var_data.py [hunk 1]
```diff
@@ -24,13 +25,23 @@ def _setup(self):
...
+        dimension_names = list(dimensions_dict.keys())
         cf_data = self.mocker.MagicMock(
             _FillValue=None,
+            _Encoding="",
             __getitem__="<real-data>",
-            dimensions=["dim_" + str(x) for x in range(len(shape))],
+            dimensions=dimension_names,
+            # A parent .group() with dimensions is needed to validate the var dims
+            group=self.mocker.Mock(
+                return_value=self.mocker.Mock(dimensions=dimensions_dict)
```
Comment (stephenworsley) [inline lib/iris/tests/unit/fileformats/netcdf/loader/test__get_cf_var_data.py]: It would make sense for the above two instances of `self.mocker.Mock` should be replaced with the variable `mock` defined on line 28.

File: lib/iris/tests/unit/fileformats/netcdf/test_bytecoding_datasets.py [hunk 2]
```diff
@@ -553,3 +561,118 @@ def test_auto_chartostring(self, samplefile_path, classtype, component_type, is_
         else:
             # Just check method exists +  doesn't error.
             component.set_auto_chartostring(is_on)
+
+
+# specific tests for underlying support classes
+# (not strictly public, but possible for use by ncdata ?
+class TestEncodeDecodeFuncs:
+    # TODO: make this work with nonstandard chuhks (currently does NOT)
```
Comment (stephenworsley) [inline lib/iris/tests/unit/fileformats/netcdf/test_bytecoding_datasets.py]: typo in "chunks"

File: lib/iris/tests/unit/fileformats/netcdf/test_bytecoding_datasets.py [hunk 3]
```diff
@@ -553,3 +561,118 @@ def test_auto_chartostring(self, samplefile_path, classtype, component_type, is_
...
+    @pytest.fixture(params=["utf8", "ascii"])
+    def encoding(self, request):
+        return request.param
+
+    @pytest.fixture(params=["lazy", "concrete"])
+    def lazyreal(self, request):
+        return request.param
+
+    SAMPLE_STRINGS_UNICODE = [["1", "éclair", "two"]] * 2
+    SAMPLE_STRINGS_ASCII = [["1", "bun", "London"]] * 2
```
Comment (stephenworsley) [inline lib/iris/tests/unit/fileformats/netcdf/test_bytecoding_datasets.py]: Is it worth also testing with empty strings?

File: lib/iris/tests/unit/fileformats/netcdf/loader/test__get_cf_var_data.py (hunk 1 above)
Comment (stephenworsley) [inline lib/iris/tests/unit/fileformats/netcdf/loader/test__get_cf_var_data.py:41]: It would make sense to also replace the above two lines with the `mock` variable defined on line 28

**Comment by pp-mo** [comment]
Rebased + added changelog fragment.

File: lib/iris/fileformats/netcdf/_bytecoding_datasets.py [hunk 4]
```diff
@@ -189,26 +194,66 @@ def _get_string_width(self) -> int:
...
             encoding = self.read_encoding
             strlen = self.string_width
-            data = decode_bytesarray_to_stringarray(
-                data, encoding, strlen, self.varname
-            )
-
-        return data
+            if not data.shape:
+                # If data is scalar, give it a dim for the operation to 'remove'.
+                # I.E. equivalent to a final dimension "string1 = 1"
```
Comment (stephenworsley) [inline lib/iris/fileformats/netcdf/_bytecoding_datasets.py]: I'm not sure I entirely understand this comment. Is this suggesting that this is how the dimension appears on the cube once it is loaded? I think this could be more specific about what "final" means in this context.

**Review by stephenworsley — CHANGES_REQUESTED** [review]
Looking good, just a couple minor things to clear up.

File: lib/iris/fileformats/netcdf/_bytecoding_datasets.py (hunk 4 above)
Comment (pp-mo) [reply lib/iris/fileformats/netcdf/_bytecoding_datasets.py]: No, I just meant an additional _last_ dimension of the array.
Will improve..

**Review by stephenworsley — APPROVED** [review]
Looks good!
